src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs

```diff
@@ -329,41 +326,103 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
         }
     }
 
-    internal sealed class ReadableStreamPullState
+    internal sealed class WasmHttpWriteStream : Stream
     {
-        private readonly Stream _stream;
-        private readonly CancellationToken _cancellationToken;
-        private readonly byte[] _buffer;
+        private readonly JSObject _transformStream;
 
-        public ReadableStreamPullState(Stream stream, CancellationToken cancellationToken)
+        public WasmHttpWriteStream(JSObject transformStream)
         {
-            ArgumentNullException.ThrowIfNull(stream);
+            ArgumentNullException.ThrowIfNull(transformStream);
 
-            _stream = stream;
-            _cancellationToken = cancellationToken;
-            _buffer = new byte[65536];
+            _transformStream = transformStream;
         }
 
-        public async Task PullAsync()
+        public static async Task<JSObject> CopyToAsync(Task<JSObject> fetchResponsePromise, JSObject transformStream, HttpContent content, CancellationToken cancellationToken)
         {
-            try
-            {
-                int length = await _stream.ReadAsync(_buffer, _cancellationToken).ConfigureAwait(true);
-                ReadableStreamControllerEnqueueUnsafe(this, _buffer, length);
-            }
-            catch (Exception ex)
+            using (WasmHttpWriteStream stream = new WasmHttpWriteStream(transformStream))
             {
-                BrowserHttpInterop.ReadableStreamControllerError(this, ex);
+                try
+                {
+                    Task copyTask = content.CopyToAsync(stream, cancellationToken);
+                    if (await Task.WhenAny(copyTask, fetchResponsePromise).ConfigureAwait(true) == copyTask)
```

Comment:
Race with fetch to get errors and cancellation because TransformStream does not propagate these.

Reply:
I wonder if race on TS side could propagate the failure to copyTask, so that we don't have to race here.

Reply:
It depends on what the behavior should be when a request content just waits indefinitely in `SerializeToStreamAsync`. E.g. should it throw a cancellation exception or not.

I checked the normal `SocketsHttpHandler` and it waits indefinitely so this can be removed.

```c#
var cts = new CancellationTokenSource();
cts.CancelAfter(TimeSpan.FromSeconds(5));

using var client = new HttpClient();
client.Timeout = TimeSpan.FromSeconds(15);
await client.PostAsync("https://github.com", new MyHttpContent(), cts.Token); // this will wait indefinitely

class MyHttpContent : HttpContent
{
    protected override Task SerializeToStreamAsync(Stream stream, TransportContext? context)
        => new TaskCompletionSource().Task;

    protected override bool TryComputeLength(out long length)
    {
        length = 0;
        return false;
    }
}
```